src/order-sortedness.c

```diff
@@ -0,0 +1,520 @@
+      groups_size_push(p_group_infos, 1);
+    }
+
+    return VCTRS_SORTEDNESS_reversed;
+  }
+
+  // Was partially in expected order. Need to sort.
+  if (count != 0) {
+    return VCTRS_SORTEDNESS_unsorted;
+  }
+
+  // Retain the original `n_groups` to be able to reset the group sizes if
+  // it turns out we don't have expected ordering
+  struct group_info* p_group_info = groups_current(p_group_infos);
+  R_xlen_t original_n_groups = p_group_info->n_groups;
+
+  R_xlen_t group_size = 1;
+
+  // Check for expected ordering - allowing ties since we don't have to
+  // reverse the ordering.
+  for (R_xlen_t i = 1; i < size; ++i) {
+    int current = p_x[i];
+
+    int cmp = int_cmp(
+      current,
+      previous,
+      direction,
+      na_order
+    );
+
+    // Not expected ordering
+    if (cmp < 0) {
+      p_group_info->n_groups = original_n_groups;
+      return VCTRS_SORTEDNESS_unsorted;
+    }
+
+    previous = current;
+
+    // Continue group run
+    if (cmp == 0) {
+      ++group_size;
+      continue;
+    }
+
+    // Expected ordering
+    groups_size_push(p_group_infos, group_size);
+    group_size = 1;
+  }
+
+  // Push final group run
+  groups_size_push(p_group_infos, group_size);
+
+  // Expected ordering
+  return VCTRS_SORTEDNESS_sorted;
+}
+
+// Very similar to `dbl_cmp()`
+static inline
+int int_cmp(int x, int y, const int direction, const int na_order) {
+  if (x == NA_INTEGER) {
+    if (y == NA_INTEGER) {
+      return 0;
+    } else {
+      return na_order;
+    }
+  }
+
+  if (y == NA_INTEGER) {
+    return -na_order;
+  }
+
+  int cmp = (x > y) - (x < y);
+
+  return cmp * direction;
+}
+
+// -----------------------------------------------------------------------------
+
+static inline int chr_cmp(SEXP x,
+                          SEXP y,
+                          const char* c_x,
+                          const char* c_y,
+                          const int direction,
+                          const int na_order);
+
+/*
+ * Check if `p_x` is in the "expected" ordering as defined by `decreasing` and
+ * `na_last`. If `p_x` is in the expected ordering, or if it is in the strictly
+ * opposite of the expected ordering (with no ties), then groups are pushed,
+ * and a `vctrs_sortedness` value is returned indicating how to finalize the
+ * order.
+ */
+enum vctrs_sortedness chr_sortedness(const SEXP* p_x,
+                                     struct group_infos* p_group_infos,
+                                     R_xlen_t size,
+                                     bool decreasing,
+                                     bool na_last,
+                                     bool check_encoding) {
+  if (size == 0) {
+    return VCTRS_SORTEDNESS_sorted;
+  }
+
+  if (size == 1) {
+    groups_size_push(p_group_infos, 1);
+    return VCTRS_SORTEDNESS_sorted;
+  }
+
+  const int direction = decreasing ? -1 : 1;
+  const int na_order = na_last ? 1 : -1;
+
+  const void* vmax = vmaxget();
```

Comment:
> This would improve performance on Windows where reencoding will be very common.

Will it really? I only reencode if the string is not ASCII and not UTF-8. What does Windows store them as? I imagined that the ASCII case will handle most strings.

Reply:
> I imagined that the ASCII case will handle most strings.

It's common to have datasets with textual data in non English languages, especially in Asian countries etc. In that case the strings are encoded in `CE_NATIVE` on Windows.